crates/subspace-networking/src/peer_info/handler.rs

```diff
@@ -0,0 +1,281 @@
+use crate::peer_info::{protocol, PeerInfo};
+use futures::future::BoxFuture;
+use futures::prelude::*;
+use libp2p::core::upgrade::{NegotiationError, ReadyUpgrade};
+use libp2p::core::UpgradeError;
+use libp2p::swarm::handler::{
+    ConnectionEvent, DialUpgradeError, FullyNegotiatedInbound, FullyNegotiatedOutbound,
+    ListenUpgradeError,
+};
+use libp2p::swarm::{
+    ConnectionHandler, ConnectionHandlerEvent, ConnectionHandlerUpgrErr, KeepAlive,
+    NegotiatedSubstream, SubstreamProtocol,
+};
+use std::error::Error;
+use std::task::{Context, Poll};
+use std::time::Duration;
+use std::{fmt, io};
+use tracing::debug;
+
+/// The configuration for peer-info protocol.
+#[derive(Debug, Clone)]
+pub struct Config {
+    /// Protocol timeout.
+    timeout: Duration,
+
+    /// Protocol name.
+    protocol_name: &'static [u8],
+}
+
+impl Config {
+    /// Creates a new [`Config`] with the following default settings:
+    ///
+    ///   * [`Config::with_timeout`] 20s
+    pub fn new(protocol_name: &'static [u8]) -> Self {
+        Self {
+            timeout: Duration::from_secs(20),
+            protocol_name,
+        }
+    }
+
+    /// Sets the protocol timeout.
+    pub fn with_timeout(mut self, d: Duration) -> Self {
+        self.timeout = d;
+        self
+    }
+}
+
+/// The successful result of processing an inbound or outbound peer info requests.
+#[derive(Debug)]
+pub enum PeerInfoSuccess {
+    Received(PeerInfo),
+    DataSent,
+}
```

Comment:
> I'm not sure separate structs will have better usage.

My thinking was that they are not the same kind of message. One is notification and the other is confirmation (right?). So they should not be in the same enum. There could be multiple types of received messages, so one enum variant there would be peer info with potentially more added in the future (let's say incremental updates, so you don't break the protocol every time in incompatible way).

> Why did warn(missing_docs) miss it? Hmm...

I think it might have been a bit confused by the fact that you have defined this struct as `pub`, but that was not correct/necessary because it is not reachable from the outside. You might as well used `pub(crate)`. It maybe it is just a bug in Clippy, those also do happen.